pulp_2to3_migration/app/migration.py

```diff
@@ -320,6 +321,67 @@ def create_repo_version(progress_rv, pulp2_repo, pulp3_remote=None):
         pulp2_repo(Pulp2Repository): a pre-migrated repository to create a repo version for
         pulp3_remote(remote): a pulp3 remote
     """
+    def detect_path_overlap(paths):
+        """
+        Check for valid POSIX paths (ie ones that aren't duplicated and don't overlap).
+
+        Overlapping paths are where one path terminates inside another (e.g. a/b and a/b/c).
+
+        NOTE: The logic is copied from pulpcore.app.files.validate_file_paths().
+
+        This function returns the first dupe or overlap it detects. We use a trie (or
+        prefix tree) to keep track of which paths we've already seen.
+
+        Args:
+            paths (iterable of str): An iterable of strings each representing a relative path
+
+        Returns:
+            str: a path which overlaps another
+        """
+        path_trie = StringTrie(separator="/")
+        for path in paths:
+            if path in path_trie:
+                # path duplicates a path already in the trie
+                return path
+
+            if path_trie.has_subtrie(path):
+                # overlap where path is 'a/b' and trie has 'a/b/c'
+                return path
+
+            prefixes = list(path_trie.prefixes(path))
+            if prefixes:
+                # overlap where path is 'a/b/c' and trie has 'a/b'
+                return path
+
+            # if there are no overlaps, add it to our trie and continue
+            path_trie[path] = True
+
+    def resolve_path_overlap(version):
+        """
+        Remove content which path overlaps some other.
+
+        If it's a duplicated path, remove the older content.
+
+        If something is absolutely wrong and we were not able to resolve conflicts,
+        repo version creation will fail later.
+
+        Paths can be overlapping because of an old Pulp 2 bug.
+        """
+        paths = ContentArtifact.objects.filter(content__pk__in=version.content).values_list(
+            "relative_path", flat=True
+        )
+        max_conflicts = len(new_version.content) - 1
+        iteration = 0
+        while True and iteration <= max_conflicts:
+            bad_path = detect_path_overlap(paths)
+            if not bad_path:
+                # no path overlaps, we are good
+                break
+
+            # find content by its relative path
+            # if duplicated - find pulp2content and check pulp2 timestamps, remove older one.
+            # otherwise remove the one found from the version
```

Comment:
I'm currently looking into how to implement this logic, and it seems somewhat painful.
Is it overkill to try to find the latest one? Should we remove just the first one?
FWIW, pulp 2 publishes the first met one I think, at least out of 2 content units with the same relative path, it picked and published the older one :/

Any suggestions are welcome, for this logic or in general.

Reply:
Maybe you could do something like the "retain old count" - partition by `relative_path`, use an expression to get the first value.

https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/models/package.py#L32-L54
https://docs.djangoproject.com/en/3.2/ref/models/database-functions/#firstvalue

Reply:
Disclaimer: I have no idea if it would work exactly like that, I don't know much about window functions

Reply:
I ended up doing just sorting and removing everything but not the most recent one from the repo version.
Thanks for the suggestions.